### PR TITLE
2/n Simplify spawn plugins: Spawn immediately

### LINKED ISSUE #10059: Simplify multiprocessing logic in DDPSpawn plugins
## Proposed refactoring or deprecation

When working on #9987 and the corresponding refactors around the spawn plugins, I realized that the logic around the multiprocessing queue and how results are handled and returned to the trainer is quite outdated and overly complicated. This logic has outlived many changes, but we never saw a way to make it simpler. With this issue I propose several steps towards a clean and easy to follow, easy to debug code path for 

1) spawning processes
2) moving results from the subprocess back to the main process
3) returning the result back to the trainer

### Motivation

There are several components in the DDPSpawn plugin around spawning processes and handling of results that are obscure and not well documented. 
On top of that, result handling bleeds into the TrainingType base class

https://github.com/PyTorchLightning/pytorch-lightning/blob/aa1540410ff55854e050ff117c2d66f22741d182/pytorch_lightning/plugins/training_type/training_type_plugin.py#L38

and also into the trainer:

https://github.com/PyTorchLightning/pytorch-lightning/blob/aa1540410ff55854e050ff117c2d66f22741d182/pytorch_lightning/trainer/trainer.py#L1123-L1125

**This is quite confusing to anyone not familiar with the peculiarities of ddp spawn. But it does not have to be that way. The situation can be drastically improved!**

#### Step 1
Remove the `self.mp_queue` attribute from the plugin. It is not required and can be locally created and used within the recently introduced `DDPSpawnPlugin.spawn` method #10018

#### Step 2
Instead of adding items like last_path, best_path, or results to the queue one by one, add all data at once as one result tuple to the queue. 

This logic

https://github.com/PyTorchLightning/pytorch-lightning/blob/aa1540410ff55854e050ff117c2d66f22741d182/pytorch_lightning/plugins/training_type/ddp_spawn.py#L224-L238

becomes 

```py

# inside the spawned processes
def new_process(self, ...):
    # ...
    return best_path, last_path, results

# in spawn wrapper:
def _wrapped_function(...):
    result = function(*args, **kwargs)
    if self.is_global_zero:
        queue.put(move_data_to_device(result, "cpu"))  # the tuple of data

# in main process:
def spawn(self, ...):
    queue = SimpleQueue()
    mp.spawn(self._wrapped_function, args=(function, args, kwargs, mp_queue), nprocs=self.num_processes)
    return mp_queue.get()  # the tuple of data
```

This allows us to standardize and limit the queue to a single `put()` and correspondingly a single `get`. This is less error prone and easier to understand for everyone working with custom plugins. The only really complicated part where a learning curve is steep for the reader is this code snippet above. Everything else will extremely simplify.

#### Step 3
With 1) and 2) in place, we can directly return the results from the spawned function instead of caching it in the attribute `self._results`. 

#### Step 4
Finally, we can get rid of dispatch and post dispatch 
   
https://github.com/PyTorchLightning/pytorch-lightning/blob/aa1540410ff55854e050ff117c2d66f22741d182/pytorch_lightning/trainer/trainer.py#L1102-L1107

and combine it into a single `plugin.run` call or alike:

```python
self.training_type_plugin.run(self.train)  # or whatever other trainer method
```
This then cleanly generalizes across all plugins. The confusing concept of dispatch and post dispatch is gone.

#### Step 5

Proposed by @ananthsub, the next step would be to directly spawn processes with the Trainer.fit() call. This how we do it in Lite as well: 

https://github.com/PyTorchLightning/pytorch-lightning/blob/412d507a73c79f5e4f7ef14289cefe2eb2af94a7/pytorch_lightning/lite/lite.py#L387-L396

The benefits of this last step are ultimately (https://github.com/PyTorchLightning/pytorch-lightning/issues/10059#issuecomment-954444367):

- consistent execution flow between spawn and non-spawn versions since _run is shared across both
   - No longer will it be required to exclude some hooks or change hook order based on the plugin spawn type. Example: setup() hook
   - No longer regressions with Loggers being accidentally created on the main process for spawn plugins. With step 5, we will no longer need to artificially delay any logger.experiment calls.
- we can move up initialization of distributed init for spawn plugins because processes are spawned earlier. this means no subtle difference in implementations of collectives, which @four4fish found out the tedious & difficult way with Deprecate LightningDistributed and keep logic in ddp/ddpSpawn directly #9691

### Nexts steps
A quick and dirty draft is available here in form of a PR (excluding some steps): #10034

### Additional context

The `add_to_queue` and `get_from_queue` methods were recently introduced, initially on the LightningModule and now they are in a deprecation phase. We would need to incorporate them into this design as well. Suggestions welcome.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #10059 (step 4 & 5)

Moves the spawning of process right at the beginning of `Trianer.fit/validate/test/predict`.
- Greatly improve readability due to simplified code path
- Aligns the code path of DDPSpawn with other plugins like DDP
- Hooks like setup() and teardown() now consistently run inside the worker processes and process group initialized
- Significantly decreases the chance that the `root_device` gets accessed in a hook that runs outside the spawned processes in TPU spawn. This today would cause the process to exit with a cryptic message. 

The core change of is that the spawning happens a level higher up in the trainer. Here is a simplified illustration using functions for before vs. after:

Before:

```py
def _run(model, ...):
    setup()
    other()
    # <-- processes spawn here
    mp.spawn(train_fn, args=model)
    # <-- processes join here

    teardown()
```

After:

```py
def _run(model):
    setup()
    other()
    train_fn(model)
    teardown()

# <-- processes spawn here
mp.spawn(_run, args=model)
# <-- processes join here
```

Possible issues that this PR might fix or help fix (need to check):
- #10688 
-  #9876

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @justusschock @kaushikb11 @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -676,7 +676,12 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
             **kwargs: keyword arguments to be passed to `trainer_fn`
         """
         try:
-            return trainer_fn(*args, **kwargs)
+            if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+                spawn_output: _SpawnOutput = self.training_type_plugin.spawn(trainer_fn, *args, **kwargs)
+                self.training_type_plugin._recover_results_in_main_process(spawn_output, self)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:681]: I really like the simplification this offers. I am wary of tying this to DDP Spawn however. This forces everyone to subclass DDPSpawnPlugin and override different parts.

What do you think about offering an AbstractSpawn interace which requires people to implement `spawn` and `recover_results_in_main_process` and which defines the `spawn_output` struct?

Then DDP Spawn could both inherit from TrainingTypePlugin and implement `AbstractSpawn`

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -402,7 +402,6 @@ def test_model_checkpoint_no_extraneous_invocations(tmpdir):
         max_epochs=num_epochs,
     )
     trainer.fit(model)
-    assert trainer.state.finished, f"Training failed with {trainer.state}"
```
Comment (awaelchli) [inline tests/checkpointing/test_model_checkpoint.py:405]: Here is one blocker I am facing, in short: 

```python
trainer = Trainer(strategy="ddp_spawn", devices=2) 
trainer.fit(model)
assert trainer.state == TrainerState.finished
```

The main process spawns immediately 2 processes with the fit call (as proposed by this new design). These workers will update the state in their own instance of the trainer but once processes terminate/join, these changes will not be reflected in the main process, which has never seen any updates (as expected). 


A solution to this would be to put the trainer.state variable onto the queue and restore it in the main process. I am wondering if this is worth it. RFC from reviewers.

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:681]: @ananthsub I was actually thinking in a similar direction, I imagined it with composition rather than inheritance. My idea was to have an (abstract) Executor interface, of which the TTP owns an instance like the checkpoint IO plugin. 
I'm definitely open to discuss such directions in more detail

Comment (justusschock) [reply pytorch_lightning/trainer/trainer.py:680]: @awaelchli You mean something like an internal launcher? That either calls the script or uses spawn? I think we should do that to make it way more modular.

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:680]: Yes I was thinking something like that.

File: tests/checkpointing/test_model_checkpoint.py (hunk 2 above)
Comment (justusschock) [reply tests/checkpointing/test_model_checkpoint.py:405]: Why don't we spawn N-1 processes and use the main proc as well?

Other then that we could probably send it through the queue as that should be quite cheap to do.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 3]
```diff
@@ -377,13 +350,3 @@ def checkpoint_io(self) -> CheckpointIO:
     @checkpoint_io.setter
     def checkpoint_io(self, plugin: CheckpointIO) -> None:
         raise MisconfigurationException("TPU Spawn Plugin currently does not support custom checkpoint plugins.")
-
-    @staticmethod
-    def _clean_logger(trainer: "pl.Trainer") -> None:
-        loggers = trainer.logger._logger_iterable if isinstance(trainer.logger, LoggerCollection) else [trainer.logger]
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:383]: Is it still necessary? It could be if the logger is still pickled right?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/trainer.py:679]: Will you open an issue for tracking this? :)

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1113,25 +1124,19 @@ def _run(
              Lightning internal flow looks like this:
         {Trainer.fit} or {Trainer.test} or {Trainer.predict}  ||
                                 |                             ||
-                        create accelerator                    ||
-                                |                             ||
-                         {self._dispatch}                     ||
+                        setup accelerator                     ||
+                           and strategy                       ||
```
Comment (justusschock) [inline pytorch_lightning/trainer/trainer.py:1128]: should we mention that this may include spawning processes?

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -62,6 +62,10 @@
 if _OMEGACONF_AVAILABLE:
     from omegaconf import OmegaConf
 
+ProcessRaisedException = Exception
+if _TORCH_GREATER_EQUAL_1_8:
+    from torch.multiprocessing import ProcessRaisedException
```
Comment (justusschock) [inline tests/trainer/test_trainer.py:67]:
```suggestion
if _TORCH_GREATER_EQUAL_1_8:
    from torch.multiprocessing import ProcessRaisedException
else:
    ProcessRaisedException = Exception
```

just a nit, but doesn't make sense to have a value and immediately overwrite it :)

**Review by justusschock — COMMENTED** [review]
Will the processes live for the entire lifetime of the trainer or just for the duration of fit? 

I.e. when I first call fit and then test, will it use the same processes or will new processes be spawned for testing?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 6]
```diff
@@ -118,10 +118,23 @@ def connect(self, model: "pl.LightningModule") -> None:
         return super().connect(model)
 
     def pre_dispatch(self, trainer: "pl.Trainer") -> None:
-        super().pre_dispatch(trainer)
+        self._move_optimizer_state()
         if self.debug:
             os.environ["PT_XLA_DEBUG"] = str(1)
 
+        if self.tpu_global_core_rank != 0 and trainer.progress_bar_callback is not None:
+            trainer.progress_bar_callback.disable()
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: Yup @awaelchli It could be removed.

**Comment by awaelchli** [comment]
@justusschock The latter. This remains the same as before. The only difference is that the spawning happens a level higher up in the trainer. I updated the PR description with a visualization.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/sharded_spawn.py [hunk 7]
```diff
@@ -114,12 +114,12 @@ def pre_backward(self, closure_loss: torch.Tensor) -> None:
     def post_training_step(self):
         pass
 
-    def new_process(self, trainer: "pl.Trainer") -> Optional["_SpawnOutput"]:
+    def pre_dispatch(self, trainer: "pl.Trainer") -> None:
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/sharded_spawn.py:117]: Is this override needed now? If not this could be cleanup for a separate PR

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -1158,10 +1163,8 @@ def _run(
 
         self.checkpoint_connector.resume_end()
 
-        # dispatch `start_training` or `start_evaluating` or `start_predicting`
-        results = self._dispatch()
-
-        self._post_dispatch()
+        results = self.run_stage()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1166]: nit: run_stage could be marked private as it's not used outside the class. For another PR 😀

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: super nit for a separate issue/PR: are external callers expected to use tpu_global_core_rank or tpu_local_core_rank? why not use self.global_rank ?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 9]
```diff
@@ -263,6 +243,10 @@ def get_mp_spawn_kwargs(self, trainer: Optional["pl.Trainer"] = None) -> Dict[st
         }
 
     def spawn(self, function: Callable, *args: Any, **kwargs: Any) -> Optional[Union[Any, "_SpawnOutput"]]:
+        # TODO: this todo is unclear, does it still apply?
+        # todo: precision pluging is call in accelerator setup and should be moved
+        if "XLA_USE_BF16" in os.environ:
+            del os.environ["XLA_USE_BF16"]
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:249]: n00b question: why is this key removed? 

before this was removed only before training, but not evaluation/prediction. was that a bug in the previous code that this is fixing?

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -721,6 +727,7 @@ def fit(
 
             datamodule: An instance of :class:`~pytorch_lightning.core.datamodule.LightningDataModule`.
         """
+        self.training_type_plugin.model = model
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:730]: n00b question: why model vs lightning_module?

File: tests/trainer/test_trainer.py [hunk 11]
```diff
@@ -1411,7 +1415,7 @@ def predict(
         callbacks=[cb, cb_1] if use_callbacks else [],
     )
     if strategy == "ddp_spawn":
-        with pytest.raises(MisconfigurationException):
+        with pytest.raises(ProcessRaisedException, match="`return_predictions` should be set to `False`"):
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply tests/trainer/test_trainer.py:1419]: this should be mentioned in the changelog in case people are catching specific exceptions from lightning

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:679]: Yes, this is quite exciting. This would enable us to possibly merge ddp, ddp_spawn into ddp.py and provide a popen or spawn executor.

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:730]: I believe this was to differentiate with `self.lightning_module`, but it might be confusing to the readers.

File: pytorch_lightning/trainer/trainer.py [hunk 12]
```diff
@@ -1223,25 +1226,18 @@ def _log_hyperparams(self) -> None:
...
-    def _dispatch(self) -> Any:
-        if self.evaluating:
-            return self.training_type_plugin.start_evaluating(self)
-        elif self.predicting:
-            return self.training_type_plugin.start_predicting(self)
-        else:
-            return self.training_type_plugin.start_training(self)
-
     def run_stage(self):
+        self.training_type_plugin.barrier("run-stage")
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1240]: IMO, I would move this logic within the run function or its own method. More calls might need to be moved in dispatch.

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 13]
```diff
@@ -132,23 +132,6 @@ def set_world_ranks(self, process_idx: int = 0) -> None:
     def get_mp_spawn_kwargs(self, trainer: Optional["pl.Trainer"] = None) -> Dict[str, Any]:
         return {"nprocs": self.num_processes}
 
-    def start_training(self, trainer: "pl.Trainer") -> Any:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:135]: This is aw !

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:679]: Ideally, we would have `self.training_type_plugin.run(trainer_fn, *args, **kwargs)` and this would execute the associated trainer method with the spawn / none executor.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1128]: Yes, I would prefer so.

File: pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -1171,15 +1174,15 @@ def _run(
...
-        # been included at the end of `new_process` functions)
-        if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
-            self._call_teardown_hook()
+        self._call_teardown_hook()
 
         if self.state.status != TrainerStatus.INTERRUPTED:
             self.state.status = TrainerStatus.FINISHED
         self.state.stage = None
 
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1183]: Any chance we can prevent strategy leakage here ?

File: pytorch_lightning/trainer/trainer.py [hunk 15]
```diff
@@ -1223,25 +1226,18 @@ def _log_hyperparams(self) -> None:
...
-            return self.training_type_plugin.start_evaluating(self)
-        elif self.predicting:
-            return self.training_type_plugin.start_predicting(self)
-        else:
-            return self.training_type_plugin.start_training(self)
-
     def run_stage(self):
+        self.training_type_plugin.barrier("run-stage")
         self.training_type_plugin.dispatch(self)
         self.__setup_profiler()
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1242]: Should this be moved to its own method ?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 16]
```diff
@@ -154,37 +166,6 @@ def init_dist_connection(self, global_rank: int, world_size: int) -> None:
...
-
-        results = trainer.run_stage()
-
-        outputs = self._collect_rank_zero_results(trainer, results)
-
-        # https://github.com/pytorch/xla/issues/1801#issuecomment-602799542
-        self.barrier("end-process")
-
-        # https://github.com/pytorch/xla/issues/2190#issuecomment-641665358
-        if self.local_rank == 0:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:181]: This might be required for TPU. @kaushikb11

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: progress bar disabling can be further refactored. posted this issue to discuss further: https://github.com/PyTorchLightning/pytorch-lightning/issues/10982

File: pytorch_lightning/trainer/trainer.py [hunk 17]
```diff
@@ -1171,15 +1174,15 @@ def _run(
             self._call_callback_hooks("on_fit_end")
             self._call_lightning_module_hook("on_fit_end")
 
-        # teardown if necessary (similar calls for spawn plugins are excluded as they have
-        # been included at the end of `new_process` functions)
-        if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
-            self._call_teardown_hook()
```
Comment (four4fish) [inline pytorch_lightning/trainer/trainer.py:1177]: Where was teardown hook get called for spawning before?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 16 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:181]: You are right, and we have this exact mechanism already in the `_wrapper_function()` at the very end. We added this when we added TPU support for Lite. So all good here, rank 0 will exit last, as required :)

File: pytorch_lightning/trainer/trainer.py [hunk 18]
```diff
@@ -1171,15 +1174,15 @@ def _run(
...
+        self._call_teardown_hook()
 
         if self.state.status != TrainerStatus.INTERRUPTED:
             self.state.status = TrainerStatus.FINISHED
         self.state.stage = None
 
+        if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+            results = self.training_type_plugin._collect_rank_zero_results(self, results)
```
Comment (four4fish) [inline pytorch_lightning/trainer/trainer.py:1184]: The _collect_rank_zero_results called after teardown hook, will there be issue if processes got killed in teardown hook?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:679]: Awesome, we will continue discussion and drafting over here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10985

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:125]: Let's investigate this. https://github.com/PyTorchLightning/pytorch-lightning/issues/10986

File: pytorch_lightning/trainer/trainer.py (hunk 12 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1240]: +1 can we deprecate run_stage()?

Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1240]: created a issue to discuss further. https://github.com/PyTorchLightning/pytorch-lightning/issues/10987

File: tests/trainer/test_trainer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/trainer/test_trainer.py:1419]: good point. added this

File: pytorch_lightning/trainer/trainer.py (hunk 17 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1177]: directly in the DDPSpawnPlugin, here: https://github.com/PyTorchLightning/pytorch-lightning/blob/aeb0b5595fd73d086f4ae0f99d3f1f112f6a4c29/pytorch_lightning/plugins/training_type/ddp_spawn.py#L205

But now, this is no longer split.

Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1177]: Oh I see, it teared down in the same function. Should it be after _collect_rank_zero_results then? same as comments below

File: pytorch_lightning/trainer/trainer.py (hunk 18 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1184]: _collect_rank_zero_results needs to be at the very and, because it needs to include the latest state of trainer.state. Whether teardown runs before or after this does not matter at the moment (process killing does not happen inside teardown). We may want to collect ideas how to polish this in #10985

File: pytorch_lightning/trainer/trainer.py (hunk 14 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1183]: Not at this moment, but some ideas here #10985  might help to do that in the future.

File: pytorch_lightning/trainer/trainer.py (hunk 18 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1184]: But if the processes already teardown, will we lost state_dicts?
The old new_process logic seems collected the output first? I'm probably miss something here
https://github.com/PyTorchLightning/pytorch-lightning/blob/aeb0b5595fd73d086f4ae0f99d3f1f112f6a4c29/pytorch_lightning/plugins/training_type/ddp_spawn.py#L202-L205

File: pytorch_lightning/plugins/training_type/sharded_spawn.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/sharded_spawn.py:117]: I'm not sure yet, I need to investigate in #10988, because the ShardedScaler gets set at instantiation of the precision plugin which originally happens in the main process on Trainer init.

File: pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1166]: Good that you mention this. I looked at usages and found HorovodPlugin is the only one left that calls `trainer.run_stage()`. I completely missed this until now. Hashing it out in #10989 as a prerequisite for the PR here.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:249]: True, something is odd here. Actually, I don't think the variable should be removed at all. It is being set initially in the precision plugin connect() method, so removing it here right after makes no sense. I will address this in #10990.

File: pytorch_lightning/trainer/trainer.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1242]: I think not worth it, because dispatch() will disappear

File: pytorch_lightning/trainer/trainer.py (hunk 18 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1184]: The teardown() call does not kill the process. Teardown() is just calling accelerator.teardown() etc. and teardown() hooks on LM etc. runs there. 

Here is what happens inside _run():

```py
def _run(...):
     # each worker process runs this function
     # <--- worker process starts here

     setup()
     ...
     run_stage()
     ...
     teardown() 

     _collect_rank_zero_results()

     # end of the function, all worker processes are done and will shut down
     return
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 16 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:181]: Sounds good to me! We had this discussion during the (long & sweet) debugging for Lite haha.

Thanks for the clean up! It reflects the same expected behavior.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:248]: Ok, as discussed in #10990, I will carry this bug over in this PR and fix it by teardown() in the linked PR. Resolving

File: pytorch_lightning/trainer/trainer.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1240]: Thanks @four4fish, will follow up on it in  #11000

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:730]: `TTP.lightning_module` doesn't have a setter, `TTP.model` has a setter. 
Initially, `TTP.lightning_module == TTP.model`, and later during setup TTP.model will hold the wrapped model.
This line that I added here (and in the other Trainer methods) is needed for the case when:

```py
trainer.fit(model)
trainer.test()  # no model provided, need to reuse self.lightning_module

```
For spawn plugins, the model reference would be lost if we didn't save it in the main process.

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: updated with your suggestions

File: pytorch_lightning/trainer/trainer.py (hunk 17 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1177]: I prefer not to change the order in the Trainer._run() for existing behavior. 
It's not a big problem that teardown() is not the very very last hook to run. If we want to change the order, I believe we  should discuss the impact and do it in another PR.

**Comment by awaelchli** [comment]
I have addressed all major questions and reviews. Thanks for the input and positive response so far!

**Review by ananthsub — APPROVED** [review]
excellent! super exciting to see this simplification

**Comment by four4fish** [comment]
Looks great to me! I don't have permission to accept, but super happy to see this 💯  A big milestone for Accelerator stable version!!

**Review by kaushikb11 — APPROVED** [review]
Woohoo!

**Review by tchaton — APPROVED** [review]
Awesome work !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:383]: It might be, but only if the users doesn't something special
